### PR TITLE
Fix SemanticsNode.rect position for nested scrollables with useTwoPan…

### LINKED ISSUE #61631: a11y focus moves randomly in GridView with many children.
b/161434746

During the navigation in GridView, when GridView is scrolled by swiping, it is likely that A11y focus will jump randomly in the page. The black border indicates A11y focus will disappear sometimes. 

```dart
import 'package:flutter/material.dart';

void main() => runApp(A11yApp());

class A11yApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final appBar =
    AppBar(title: const Icon(Icons.archive));

    return MaterialApp(
      home: Scaffold(
        appBar: appBar,
        body: SingleChildScrollView(
          child: Column(
            children: [_buildGridView(), _buildGridView(), _buildGridView()],
          ),
        ),
      ),
    );
  }

  Widget _buildGridView() {
    return Column(
      children: [
        const Text('Grid header'),
        GridView.count(
          shrinkWrap: true,
          crossAxisSpacing: 10,
          mainAxisSpacing: 10,
          crossAxisCount: 2,
          children: <Widget>[
            for (var i = 0; i < 50; ++i)
              Container(
                padding: const EdgeInsets.all(8),
                color: Colors.green[100],
                child: Text('Number $i'),
              ),
          ],
        ),
      ],
    );
  }
}
```

This is blocking release due to a11y review. See internal bug for videos on the behavior.

### PR DESCRIPTION
**Pull request by nscobie**
…eSemantics

## Description

**Issue**

If a scrolling widget such as `ListView` or `GridView` (with `shrinkWrap: true`) is nested under another scrolling widget such as `SingleChildScrollView`, then new "beyond the fold" content which is revealed by swiping off screen cannot be selected by tap, and continued swiping beyond the fold will eventually make a11y focus jump to a random element on iOS.

<details>
<summary>Repro case</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(A11yApp());

class A11yApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final appBar =
    AppBar(title: const Icon(Icons.archive));

    return MaterialApp(
      home: Scaffold(
        appBar: appBar,
        body: SingleChildScrollView(
          child: Column(
            children: [_buildGridView(), _buildGridView(), _buildGridView()],
          ),
        ),
      ),
    );
  }

  Widget _buildGridView() {
    return Column(
      children: [
        const Text('Grid header'),
        GridView.count(
          shrinkWrap: true,
          crossAxisSpacing: 10,
          mainAxisSpacing: 10,
          crossAxisCount: 2,
          children: <Widget>[
            for (var i = 0; i < 50; ++i)
              Container(
                padding: const EdgeInsets.all(8),
                color: Colors.green[100],
                child: Text('Number $i'),
              ),
          ],
        ),
      ],
    );
  }
}
```

</details>

**Fix**

The inner [semantics pane](https://api.flutter.dev/flutter/rendering/RenderViewport/useTwoPaneSemantics-constant.html) of the child `ListView`/`GridView` needs to have its position updated to match the outer semantics pane when the parent `SingleChildScrollView` scrolls.

**[This comment](https://github.com/flutter/flutter/issues/61631#issuecomment-663729586) on issue #61631 explains the technical details of the fix.**

## Related Issues

Fixes #61631 

## Tests

I added the following tests:

- `widgets/scrollable_semantics_test.dart`
    - "transform of inner node from useTwoPaneSemantics scrolls correctly with nested scrollables"
        - Ensures the _global_ `rect` of the inner and outer semantics panes of a `ListView` match when nested under a `SingleChildScrollView`. The parent `SingleChildScrollView` is scrolled to the end -> beginning -> middle of the shrink-wrapped `ListView`'s _hidden_ elements to test both scroll directions.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by nscobie** [comment]
@goderbauer This behavior wasn't covered by tests before, and I don't have enough knowledge on how the framework is typically used to know if this could break other sliver/scrolling/nesting configurations. Can you think of any issues this might introduce?

The debugging logs in the [linked comment](https://github.com/flutter/flutter/issues/61631#issuecomment-663729586) might give more context on the issue, but are admittedly hard to follow. I'm available to discuss further or elaborate on the logs.

File: packages/flutter/test/widgets/scrollable_semantics_test.dart [hunk 1]
```diff
@@ -611,3 +656,13 @@ Future<void> fling(WidgetTester tester, Offset offset, int repetitions) async {
     await tester.pump(const Duration(seconds: 5));
   }
 }
+
+Rect nodeGlobalRect(SemanticsNode node) {
```
Comment (nscobie) [inline packages/flutter/test/widgets/scrollable_semantics_test.dart:660]: Is there an existing function which already handles this? If not, should I move this somewhere more centralized?

**Comment by chunhtai** [comment]
cirrus looks unhappy, can you rebase off the master branch?

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 2]
```diff
@@ -784,7 +784,7 @@ class _RenderScrollSemantics extends RenderProxyBox {
     _innerNode ??= SemanticsNode(showOnScreen: showOnScreen);
     _innerNode
       ..isMergedIntoParent = node.isPartOfNodeMerging
-      ..rect = Offset.zero & node.rect.size;
+      ..rect = node.rect;
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/scrollable.dart:787]: This code looks correct to me, I am curious though why we want to & with the offset.zero at the first place @goderbauer

**Comment by nscobie** [comment]
> cirrus looks unhappy, can you rebase off the master branch?

Had to fix some lints I missed, but seems good now.

**Review by gaaclarke — COMMENTED** [review]
The code and test looks good to me.  I'll defer to the others with respect to the wisdom of the change as-is.  Nice sleuthing.

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/scrollable.dart:787]: I think we forgot to change this when the code was moved in https://github.com/flutter/flutter/pull/13983 a loong time ago!

File: packages/flutter/test/widgets/scrollable_semantics_test.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/test/widgets/scrollable_semantics_test.dart:660]: I don't think this is needed very often. Usually you only care about their relative position to the parent. Having that just here seems fine.

**Review by goderbauer — APPROVED** [review]
LGTM
